Tools/wasm/wasm_build.py

```diff
@@ -0,0 +1,478 @@
+#!/usr/bin/env python3
+"""Build script for Python on WebAssembly platforms
+
+  $ ./Tools/wasm/wasm_builder.py emscripten-browser compile
+  $ ./Tools/wasm/wasm_builder.py emscripten-node-dl test
+  $ ./Tools/wasm/wasm_builder.py wasi test
+
+Primary build targets are "emscripten-node-dl" (NodeJS, dynamic linking),
+"emscripten-browser", and "wasi".
+
+Emscripten builds require Emscripten SDK. The tools looks for 'EMSCRIPTEN'
+env var and falls back to EMSDK installed at /opt/emsdk.
+
+WASI builds require WASI SDK and wasmtime. The tool looks for 'WASI_SDK_PATH'
+and falls back to /opt/wasi-sdk.
+"""
+import argparse
+import enum
+import dataclasses
+import os
+import pathlib
+import shlex
+import shutil
+import subprocess
+import sysconfig
+from typing import Callable, List, Union  # for Python 3.8
+
+SRCDIR = pathlib.Path(__file__).parent.parent.parent.absolute()
+WASMTOOLS = SRCDIR / "Tools" / "wasm"
+BUILDDIR = SRCDIR / "builddir"
+CONFIGURE = SRCDIR / "configure"
+SETUP_LOCAL = SRCDIR / "Modules" / "Setup.local"
+
+HAS_CCACHE = shutil.which("ccache") is not None
+
+# path to WASI-SDK root
+WASI_SDK_PATH = pathlib.Path(os.environ.get("WASI_SDK_PATH", "/opt/wasi-sdk"))
+# path to Emscripten directory (upstream/emscripten subdirectory in the EMSDK)
+EMSCRIPTEN_PATH = pathlib.Path(os.environ.get("EMSCRIPTEN", "/opt/emsdk/upstream/emscripten"))
+
+# WASM_WEBSERVER = WASMTOOLS / "wasmwebserver.py"
+
+INSTALL_EMSDK = """
+wasm32-emscripten builds need Emscripten SDK. Please follow instructions at
+https://emscripten.org/docs/getting_started/downloads.html how to install
+Emscripten tp "/opt/emsdk". Alternatively you can install the SDK in a
```

Comment:
The hard-coded fallback to `/opt/emsdk` is a trick to make my life a bit easier. It is possible to use the SDK without `source emsdk_env.sh`. The `emconfigure` and `emmake` scripts detect the correct paths even when the SDK is not activated. I'm totally fine to keep this as a hidden feature.

Reply:
Makes sense.  In that case, the least surprising behavior is probably to look up in `PATH`, and fallback to `/opt/emsdk`.  That will work with `source emsdk_env.sh` if people are using it, also work with system packages of emscripten [1], and not require setting another environment variable (`EMSCRIPTEN`).

[1] At least on Debian, the system package doesn't really work out-of-the-box since it can't install "ports", but someday it might.

Reply:
I should add a note that users should not use Debian/Ubuntu packages. They are out of date. Ubuntu 22.04 has a version with 32bit time_t and 22.10-dev has a broken version.